src/core/thread/mle_router.cpp

```diff
@@ -335,7 +334,52 @@ ThreadError MleRouter::HandleChildStart(otMleAttachFilter aFilter)
         break;
 
     case kMleAttachSamePartition:
-        SendAddressRelease();
+        if (mDeviceMode & ModeTlv::kModeFFD && IsRouterIdValid(mPreviousRouterId))
+        {
+            // downgrade
+            if (GetActiveRouterCount() > mRouterDowngradeThreshold)
+            {
+                SendAddressRelease();
+
+                // reset children info if any
+                if (HasChildren())
+                {
+                    for (uint8_t i = 0; i < mMaxChildrenAllowed; i++)
+                    {
+                        if (mChildren[i].mState == Neighbor::kStateRestored ||
+                            mChildren[i].mState == Neighbor::kStateChildUpdateRequest ||
+                            mChildren[i].mState == Neighbor::kStateValid)
+                        {
+                            RemoveStoredChild(mChildren[i].mValid.mRloc16);
+                        }
+
+                        mChildren[i].mState = Neighbor::kStateInvalid;
+                    }
+                }
+
+                // reset routerId info
+                SetRouterId(kInvalidRouterId);
```

Comment:
Or do you think it would be better to add `HandleAddressReleaseResponse()` to reset children info and routerId info? @jwhui